### PR TITLE
Add unused_capture_list rule

### LINKED ISSUE #2715: Rule Request: unused value in capture group
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New rule request

1. Why should this rule be added? Share links to existing discussion about what
   the community thinks about this.

Similar to unused parameter in closure.

2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

```swift
// Triggering
{ [foo] in _ }()

// Non- Triggering
{ [foo] in foo.bar() }()
```

3. Should the rule be configurable, if so what parameters should be configurable?

Only severity.

4. Should the rule be opt-in or enabled by default? Why?
   See [README.md](../README.md#opt-in-rules) for guidelines on when to mark a rule as opt-in.

Enabled by default, same as `unused_closure_parameter`.

### PR DESCRIPTION
**Pull request by daltonclaybrook**
Resolves #2715 

I'm using an `ASTRule` in order to find closures, but I was unable to find any first-class parsing support within SourceKitten for capture lists, so I am doing this with regex and string manipulation. Let me know if I missed something.

### CODE REVIEW

File: Source/SwiftLintFramework/Rules/Lint/UnusedCaptureListRule.swift [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            """
+            let failure: Failure = { [↓weak self, ↓unowned delegate = self.delegate!] foo in
+                print(foo)
+            }
+            """,
+            "{ [↓foo] in _ }()"
+        ]
+    )
+
+    private let captureListRegex = regex("^\\{\\h*\\[([^\\]]+)\\].*\\bin\\b")
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/Lint/UnusedCaptureListRule.swift]: any particular reason to use `\h` here? although not common, the capture list could be in a new line

File: Source/SwiftLintFramework/Rules/Lint/UnusedCaptureListRule.swift [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+public struct UnusedCaptureListRule: ASTRule, ConfigurationProviderRule, AutomaticTestableRule {
+    public var configuration = SeverityConfiguration(.warning)
+
+    public init() {}
+
+    public static var description = RuleDescription(
+        identifier: "unused_capture_list",
+        name: "Unused Capture List",
+        description: "Unused reference in a capture list should be removed.",
+        kind: .lint,
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/Lint/UnusedCaptureListRule.swift]: can you add `swiftVersion: .fourDotTwo`? `SwiftSyntaxKind.closure` is only returned by SourceKit in 4.2+

File: Source/SwiftLintFramework/Rules/Lint/UnusedCaptureListRule.swift [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                    .trimmingCharacters(in: .whitespacesAndNewlines)
+                    .components(separatedBy: .whitespaces)
+                    .last
+                if let reference = reference {
+                    referencesAndLocations.append((reference, location))
+                }
+            }
+    }
+
+    private func identifierStrings(in file: File, byteRange: NSRange) -> [String] {
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/Lint/UnusedCaptureListRule.swift]: nitpick:: this could return a `Set<String>`

**Comment by marcelofabri** [comment]
> I'm using an ASTRule in order to find closures, but I was unable to find any first-class parsing support within SourceKitten for capture lists

They're reported as local variables inside the closure, but you'd need to be creative to figure out if it's a captured variable (i.e. check if the location is before the `in` token location).

File: Source/SwiftLintFramework/Rules/Lint/UnusedCaptureListRule.swift (hunk 1 above)
Comment (daltonclaybrook) [reply Source/SwiftLintFramework/Rules/Lint/UnusedCaptureListRule.swift]: ah, good point.

**Comment by jpsim** [comment]
Thanks for the contribution @daltonclaybrook, and this is indeed a useful rule, although there are some false positives (3/4 of the violations reported by OSSCheck are false positives). If we can't avoid those false positives, I think we should make this rule opt-in (disabled by default).

**Comment by daltonclaybrook** [comment]
@jpsim oh whoops, I guess I missed those warnings. I may have some time this afternoon to take a look at this.

**Pull request merged**